build-logic/src/main/kotlin/publishing/maven-utils.kt

```diff
@@ -0,0 +1,107 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package publishing
+
+import org.gradle.api.DefaultTask
+import org.gradle.api.Project
+import org.gradle.api.file.DirectoryProperty
+import org.gradle.api.plugins.JavaLibraryPlugin
+import org.gradle.api.provider.ListProperty
+import org.gradle.api.tasks.CacheableTask
+import org.gradle.api.tasks.Input
+import org.gradle.api.tasks.OutputDirectory
+import org.gradle.api.tasks.SourceSetContainer
+import org.gradle.api.tasks.Sync
+import org.gradle.api.tasks.TaskAction
+import org.gradle.kotlin.dsl.provideDelegate
+
+@CacheableTask
+abstract class GeneratePomProperties : DefaultTask() {
+  @Suppress("unused") @get:Input abstract val pomInputs: ListProperty<String>
+
+  @get:OutputDirectory abstract val destinationDir: DirectoryProperty
+
+  init {
+    pomInputs.convention(listOf(project.group.toString(), project.name, project.version.toString()))
+    destinationDir.convention(project.layout.buildDirectory.dir("generated/pom-properties"))
+  }
+
+  @TaskAction
+  fun generate() {
+    val buildDir = destinationDir.get().asFile
+    buildDir.deleteRecursively()
+    val targetDir = buildDir.resolve("META-INF/maven/${project.group}/${project.name}")
+    targetDir.mkdirs()
+    targetDir
+      .resolve("pom.properties")
+      .writeText(
+        """
+      # Generated by the Apache Polaris build.
+      groupId=${project.group}
+      artifactId=${project.name}
+      version=${project.version}
+    """
+          .trimIndent()
+      )
+  }
+}
+
+/**
+ * Adds convenient, but not strictly necessary information to each generated "main" jar.
+ *
+ * This includes `pom.properties` and `pom.xml` files where Maven places those, in
+ * `META-INF/maven/group-id/artifact-id/`. Also adds the `NOTICE` and `LICENSE` files in `META-INF`,
+ * which makes it easier for license scanners.
+ */
+fun addAdditionalJarContent(project: Project): Unit =
+  project.run {
+    project.plugins.withType(JavaLibraryPlugin::class.java) {
+      val generatePomProperties =
+        tasks.register("generatePomProperties", GeneratePomProperties::class.java) {}
+
+      val additionalJarContent =
+        tasks.register("additionalJarContent", Sync::class.java) {
+          // Have to manually declare the inputs of this task here on top of the from/include below
+          inputs.files(rootProject.layout.files("LICENSE", "NOTICE"))
+          inputs.property("GAV", "${project.group}:${project.name}:${project.version}")
+          dependsOn("generatePomFileForMavenPublication")
+          from(rootProject.rootDir) {
+            include("LICENSE", "NOTICE")
```

Comment:
I'm not sure about including the root NOTICE into each jar. Binary notices may have to be different for different jars because of different dependency sets.

@jbonofre : WDYT?

Reply:
It's all we have at the moment. :shrug:

Reply:
Theorically, we should have `LICENSE`/`NOTICE` actually representing what's in the jar. Taking `LICENSE`/`NOTICE` from the root folder (e.g. source distribution) is fine.

I'm working on `LICENSE`/`NOTICE` for uber jars/distributions (see #1031 ).

So let me do a pass, but using `LICENSE`/`NOTICE` sounds reassonoable.

Reply:
I still do not see the benefit of including the root NOTICE in every jar (due to it not being 100% correct) but I'll defer to @jbonofre 's opinion.